ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java

```diff
@@ -2101,6 +2102,41 @@ public boolean validateMaterializedViewsFromRegistry(List<Table> cachedMateriali
     }
   }
 
+  private Materialization getMaterializationInvalidationInfo(MaterializedViewMetadata metadata)
+      throws TException, HiveException {
+    MaterializationSnapshot mvSnapshot = MaterializationSnapshot.fromJson(metadata.creationMetadata.getValidTxnList());
+    if (mvSnapshot.getTableSnapshots() == null || mvSnapshot.getTableSnapshots().isEmpty()) {
+      return getMSC().getMaterializationInvalidationInfo(
+          metadata.creationMetadata, conf.get(ValidTxnList.VALID_TXNS_KEY));
+    }
+
+    boolean hasDelete = false;
+    for (SourceTable sourceTable : metadata.getSourceTables()) {
+      Table table = getTable(sourceTable.getTable().getDbName(), sourceTable.getTable().getTableName());
+      HiveStorageHandler storageHandler = table.getStorageHandler();
+      if (storageHandler == null) {
+        Materialization materialization = new Materialization();
+        materialization.setSourceTablesCompacted(true);
+        return materialization;
+      }
+      Boolean b = storageHandler.hasDeleteOperation(
+          table, mvSnapshot.getTableSnapshots().get(table.getFullyQualifiedName()));
+      if (b == null) {
+        Materialization materialization = new Materialization();
+        materialization.setSourceTablesCompacted(true);
+        return materialization;
```

Comment:
shouldn't we check all of the source tables?

Reply:
In this case `null` means we can not confirm that this source table has deletes or not. This means the incremental rebuild plan can not be chosen so fall back to full rebuild.
If it can not be confirmed at least one of the source tables we fall back to full rebuild.

Reply:
👍